specs/experimental/alligator.md

```diff
@@ -0,0 +1,401 @@
+revert with an error.
+
+The `subdelegateBatched` function MUST iterate over each pair of `_delegatees` address and `_rules` subdelegation rule.
+At every iteration, the function MUST check the validity of the subdelegation rule, and migrate the delegatee address
+if it has not been migrated. The function MUST then update the subdelegation. Specifically for validation,
+the function MUST check that the `allowance` field of the rule does not exceed the total voting power of the delegator'
+if the `allowanceType` is `Absolute`.
+
+Afterwards, the `subdelegateBatched` function MUST emit a `Subdelegation` event with the given function parameters.
+
+#### `afterTokenTransfer`
+
+Updates the voting power of two addresses (`_from` and `_to`) after a token transfer. This function MUST only be callable
+by the `GovernanceToken` contract.
+
+```solidity
+function afterTokenTransfer(address _from, address _to, uint256 _amount) external
+```
+
+The `Alligator` MUST check if the `_from` or `_to` addresses have been migrated by checking the `migrated` mapping
+from its [storage](#storage). If either address has not been migrated, the `Alligator` MUST copy the delegation
+and checkpoint data from the token contract to its own state. After copying the data, the `Alligator` MUST update
+the `migrated` mapping to reflect that the address has been migrated.
+
+The `afterTokenTransfer` function MUST update the voting power of the `_from` and `_to` addresses with the same logic
+of the `GovernanceToken`.
+
+#### `migrateAccounts`
+
+Migrates the delegation state of given accounts from the `GovernanceToken` to the `Alligator` contract. This function
+MUST iterate over the list of `_accounts` addresses and only migrate the account if it has not been migrated yet.
+
+```solidity
+function migrateAccounts(address[] calldata _accounts) external
+```
+
+### Getters
+
+For backwards compatibility, the `Alligator` MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been been migrated to the `Alligator` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `Alligator` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegatee with the highest voting power for a given user address. This function
+is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
```

Comment:
Why is this `delegates()(address)` rather than `delegates()([]address)`? Given that it is plural

Reply:
yes, good catch

Reply:
Original function doesn't return an array - if this is meant to replace the original function then this is correct. It's just a quirk of the naming of `ERC20Votes`.